modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java

```diff
@@ -625,6 +761,117 @@ public synchronized int read() throws IOException {
     }
 
+    private static InputStream toSynchronizedInputStream(String blobName, InputStream delegate, MultiPart multipart) {
+        assert delegate.markSupported() : "An InputStream with mark support was expected";
+        // We need to introduce a read barrier in order to provide visibility for the underlying
+        // input stream state as the input stream can be read from different threads.
+        return new FilterInputStream(delegate) {
+
+            private final boolean isTraceEnabled = logger.isTraceEnabled();
+
+            @Override
+            public synchronized int read(byte[] b, int off, int len) throws IOException {
+                var result = super.read(b, off, len);
+                if (isTraceEnabled) {
+                    logger.trace("{} reads {} bytes from {} part {}", Thread.currentThread(), result, blobName, multipart.part());
+                }
+                return result;
+            }
+
+            @Override
+            public synchronized int read() throws IOException {
+                var result = super.read();
+                if (isTraceEnabled) {
+                    logger.trace("{} reads {} byte from {} part {}", Thread.currentThread(), result, blobName, multipart.part());
+                }
+                return result;
+            }
+
+            @Override
+            public synchronized void mark(int readlimit) {
+                if (isTraceEnabled) {
+                    logger.trace("{} marks stream {} part {}", Thread.currentThread(), blobName, multipart.part());
+                }
+                super.mark(readlimit);
+            }
+
+            @Override
+            public synchronized void reset() throws IOException {
+                if (isTraceEnabled) {
+                    logger.trace("{} resets stream {} part {}", Thread.currentThread(), blobName, multipart.part());
+                }
+                super.reset();
+            }
+
+            @Override
+            public synchronized void close() throws IOException {
+                if (isTraceEnabled) {
+                    logger.trace("{} closes stream {} part {}", Thread.currentThread(), blobName, multipart.part());
+                }
+                super.close();
+            }
+
+            @Override
+            public String toString() {
+                return blobName + " part [" + multipart.part() + "] of size [" + multipart.blockSize() + ']';
+            }
+        };
+    }
+
+    private static Flux<ByteBuffer> toFlux(InputStream stream, long length, int chunkSize) {
+        assert stream.markSupported() : "An InputStream with mark support was expected";
+        // We need to mark the InputStream as it's possible that we need to retry for the same chunk
+        stream.mark(Integer.MAX_VALUE);
+        final var bytesRead = new AtomicLong(0L);
+        return Flux.defer(() -> {
+            // Code in this Flux.defer() can be concurrently executed by multiple threads
```

Comment:
Can it? I think that would break everything. I think it is an existing assumption and code comment copy - but if you agree that it should not perhaps worth adding a question mark/todo around it?

Reply:
This was the case when I worked on this last week, I could see different `repository_azure` threads resetting the same input stream instance concurrently so I assumed the existing comment was also applicable.

I updated the comment so that we can revisit this later, now I fixed the bugs I made.